File: android/app/src/main/java/dev/pepotech/pepomote/net/ScreenClient.kt
package dev.pepotech.pepomote.net

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.Socket
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.roundToInt

/** Decodifica la carga JPEG de una trama de imagen. En la app, [BitmapDecoder]. */
interface ScreenDecoder<I : Any> {
    /**
     * Imagen decodificada de `data[off, off + len)`, o null si no vale (se
     * confirma igualmente al receptor para no atascar el flujo).
     */
    fun decode(data: ByteArray, off: Int, len: Int): I?
}

/**
 * Canal de pantalla del GamePad (PROTOCOL.md §4.4): conexión TCP propia al
 * puerto PMP del receptor, línea `screen` con la sesión y el tamaño máximo
 * que cabe en la zona táctil, y de vuelta tramas [ScreenFrames] con la
 * pantalla del GamePad de Cemu en JPEG. Todo en un hilo propio: ahí se
 * decodifica (reutilizando bitmaps) y se publica en [image]; la UI avisa con
 * [shown] al pintarla y entonces sale el byte de confirmación (una imagen en
 * vuelo: la latencia no se acumula). Si nadie la pinta en [shownFallbackMs],
 * se confirma igualmente. El texto de estado del receptor (tipo 2) va en
 * [status] y deja de haber imagen; el keepalive (tipo 3) solo cuenta como
 * dato recibido.
 *
 * - Sin respuesta al `screen` en 3 s (o cierre sin contestar, o `err`):
 *   «sin pantalla» (receptor sin canal de pantalla); reintento cada 5 s.
 * - Sin datos en 6 s tras el ok: conexión muerta; se cierra y se reconecta
 *   al segundo, y si tampoco, cada 5 s.
 * - [close] es definitivo: nada vuelve a abrirse ni a publicarse.
 */
class ScreenClient<I : Any>(
    private val host: String,
    private val port: Int,
    private val sessionId: Int,
    private val maxWidth: Int,
    private val maxHeight: Int,
    private val decoder: ScreenDecoder<I>,
    private val quality: Int = QUALITY,
    private val replyTimeoutMs: Int = REPLY_TIMEOUT_MS,
    private val dataTimeoutMs: Int = DATA_TIMEOUT_MS,
    private val shownFallbackMs: Long = SHOWN_FALLBACK_MS,
    private val retryAfterDropMs: Long = RETRY_AFTER_DROP_MS,
    private val retryUnavailableMs: Long = RETRY_UNAVAILABLE_MS
) {
    companion object {
        /** Resolución nativa de la pantalla del GamePad: nunca se pide más. */
        const val NATIVE_WIDTH = 854
        const val NATIVE_HEIGHT = 480
        const val QUALITY = 70

        const val CONNECT_TIMEOUT_MS = 4000
        const val REPLY_TIMEOUT_MS = 3000
        const val DATA_TIMEOUT_MS = 6000

        /** Si la UI no avisa de que pintó la imagen en este tiempo, se confirma igualmente. */
        const val SHOWN_FALLBACK_MS = 200L
        const val RETRY_AFTER_DROP_MS = 1000L
        const val RETRY_UNAVAILABLE_MS = 5000L

        /** La respuesta es una línea corta; más que esto sin `\n` no es un receptor. */
        private const val MAX_LINE = 4096

        const val STATUS_UNAVAILABLE = "Sin pantalla (el PC no la envía)"
        const val STATUS_RECONNECTING = "Reconectando la pantalla…"

        /** Línea `screen` (PROTOCOL.md §4.4); `session_id` va como u32. */
        internal fun helloLine(sessionId: Int, w: Int, h: Int, q: Int): String =
            "{\"m\":\"screen\",\"session_id\":${sessionId.toLong() and 0xFFFF_FFFFL},\"w\":$w,\"h\":$h,\"q\":$q}"

        /** `{"m":"screen","ok":true}`: el receptor acepta. */
        internal fun replyOk(line: String): Boolean =
            jsonString(line, "m") == "screen" && Regex("\"ok\"\\s*:\\s*true").containsMatchIn(line)

        /**
         * Valor de una clave string de la línea JSON de respuesta (solo tiene
         * `m`, `ok`, `code` y `msg`). Sin org.json: el de Android no existe
         * en la JVM de los tests, y para esto no hace falta más.
         */
        internal fun jsonString(line: String, key: String): String? {
            val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"").find(line)
                ?: return null
            val raw = m.groupValues[1]
            if ('\\' !in raw) return raw
            val sb = StringBuilder(raw.length)
            var i = 0
            while (i < raw.length) {
                val c = raw[i]
                if (c != '\\' || i + 1 >= raw.length) {
                    sb.append(c)
                    i++
                    continue
                }
                when (val e = raw[i + 1]) {
                    'n' -> sb.append('\n')
                    't' -> sb.append('\t')
                    'r' -> sb.append('\r')
                    'u' -> {
                        val code = if (i + 5 < raw.length) raw.substring(i + 2, i + 6).toIntOrNull(16) else null
                        if (code != null) {
                            sb.append(code.toChar())
                            i += 4
                        }
                    }

                    else -> sb.append(e) // \" \\ \/
                }
                i += 2
            }
            return sb.toString()
        }
    }

    /** Imagen decodificada lista para pintar; [seq] la identifica (el bitmap de debajo se reutiliza). */
    class Image<I>(val seq: Int, val bitmap: I)

    private val _image = MutableStateFlow<Image<I>?>(null)
    val image: StateFlow<Image<I>?> = _image

    /** Última línea de estado (del receptor, o propia: sin pantalla / reconectando). */
    private val _status = MutableStateFlow<String?>(null)
    val status: StateFlow<String?> = _status

    /** Imágenes por segundo (media de 1 s); 0 sin imagen. */
    private val _fps = MutableStateFlow(0)
    val fps: StateFlow<Int> = _fps

    @Volatile
    private var running = true

    @Volatile
    private var socket: Socket? = null

    @Volatile
    private var out: OutputStream? = null

    /** Generación de la conexión: una confirmación de la anterior no se escribe en la nueva. */
    @Volatile
    private var conn = 0

    /** `seq` de la imagen publicada y aún sin confirmar (0 = ninguna). */
    private val pendingAck = AtomicInteger(0)
    private var seq = 0
    private val frames = ScreenFrames()
    private var fpsWindowStartNs = 0L
    private var fpsCount = 0

    /** TODA escritura al socket (las confirmaciones) pasa por este hilo. */
    private val outbound: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor { r ->
        Thread(r, "pepomote-screen-out").apply { isDaemon = true }
    }

    private val thread = Thread(::run, "pepomote-screen")

    fun start() {
        thread.start()
    }

    /** La UI acaba de pintar la imagen [seq]: ahora sí, confirmación al receptor. */
    fun shown(seq: Int) {
        if (pendingAck.compareAndSet(seq, 0)) ack(conn)
    }

    fun close() {
        running = false
        outbound.shutdownNow()
        try {
            socket?.close()
        } catch (_: Exception) {
        }
        thread.interrupt()
        _image.value = null
        _status.value = null
        _fps.value = 0
    }

    private fun run() {
        try {
            while (running) {
                val gotOk = session()
                if (!running) break
                Thread.sleep(if (gotOk) retryAfterDropMs else retryUnavailableMs)
            }
        } catch (_: InterruptedException) {
        }
    }

    /** Una conexión, hasta que se cierra. Devuelve si el receptor llegó a contestar ok. */
    private fun session(): Boolean {
        var gotOk = false
        val s = Socket()
        socket = s
        try {
            if (!running) return false
            s.tcpNoDelay = true
            s.connect(InetSocketAddress(host, port), CONNECT_TIMEOUT_MS)
            s.soTimeout = replyTimeoutMs
            val input = s.getInputStream()
            val output = s.getOutputStream()
            output.write((helloLine(sessionId, maxWidth, maxHeight, quality) + "\n").toByteArray(Charsets.UTF_8))
            output.flush()
            frames.reset()

            // Respuesta: una línea JSON. Un receptor sin pantalla cierra o calla
            // (soTimeout) y un `err` trae el motivo.
            val line = readReply(input) ?: run {
                unavailable(null)
                return false
            }
            if (!replyOk(line)) {
                unavailable(jsonString(line, "msg"))
                return false
            }
            gotOk = true
            conn++
            pendingAck.set(0)
            out = output
            s.soTimeout = dataTimeoutMs
            _status.value = null

            while (running) {
                val f = frames.next()
                if (f == null) {
                    if (frames.fill(input) < 0) break
                    continue
                }
                when (f.type) {
                    ScreenFrames.TYPE_IMAGE -> onImage(f)
                    ScreenFrames.TYPE_STATUS -> onStatus(String(f.data, f.offset, f.length, Charsets.UTF_8))
                    else -> Unit // keepalive (ya contó como dato al llegar) o tipo futuro
                }
            }
        } catch (_: Exception) {
            // Sin respuesta en 3 s, sin datos en 6 s, conexión rechazada o
            // cerrada: según la fase, «sin pantalla» o «reconectando»
            if (!gotOk) unavailable(null)
        } finally {
            out = null
            try {
                s.close()
            } catch (_: Exception) {
            }
            if (gotOk) dropped()
        }
        return gotOk
    }

    /** Línea de respuesta al `screen`, o null si el receptor cierra o no manda una línea. */
    private fun readReply(input: InputStream): String? {
        while (true) {
            frames.takeLine()?.let { return it }
            if (frames.pending > MAX_LINE || frames.fill(input) < 0) return null
        }
    }

    private fun onImage(f: ScreenFrames.Frame) {
        val bitmap = try {
            decoder.decode(f.data, f.offset, f.length)
        } catch (_: Exception) {
            null
        }
        if (!running) return
        val c = conn
        if (bitmap == null) {
            ack(c) // nada que enseñar: se confirma igual para que siga el flujo
            return
        }
        val n = ++seq
        pendingAck.set(n)
        countImage()
        _image.value = Image(n, bitmap)
        // Si la UI no llega a pintarla, que no se atasque el flujo
        try {
            outbound.schedule(
                Runnable { if (pendingAck.compareAndSet(n, 0)) writeAck(c) },
                shownFallbackMs, TimeUnit.MILLISECONDS
            )
        } catch (_: Exception) {
            // ejecutor ya cerrado
        }
    }

    /** Estado del receptor: solo lo manda cuando no tiene imagen que dar. */
    private fun onStatus(text: String) {
        _image.value = null
        resetFps()
        _status.value = text
    }

    private fun unavailable(msg: String?) {
        if (!running) return
        _image.value = null
        resetFps()
        _status.value = if (msg.isNullOrBlank()) STATUS_UNAVAILABLE else "Sin pantalla: $msg"
    }

    private fun dropped() {
        if (!running) return
        _image.value = null
        resetFps()
        _status.value = STATUS_RECONNECTING
    }

    private fun ack(c: Int) {
        try {
            outbound.execute { writeAck(c) }
        } catch (_: Exception) {
            // ejecutor ya cerrado
        }
    }

    /** Solo en el hilo outbound. */
    private fun writeAck(c: Int) {
        if (c != conn) return
        try {
            out?.apply {
                write(ScreenFrames.ACK.toInt())
                flush()
            }
        } catch (_: Exception) {
        }
    }

    private fun countImage() {
        val now = System.nanoTime()
        if (fpsWindowStartNs == 0L) fpsWindowStartNs = now
        fpsCount++
        val elapsed = now - fpsWindowStartNs
        if (elapsed >= 1_000_000_000L) {
            _fps.value = (fpsCount * 1e9 / elapsed).roundToInt()
            fpsWindowStartNs = now
            fpsCount = 0
        }
    }

    private fun resetFps() {
        fpsWindowStartNs = 0L
        fpsCount = 0
        _fps.value = 0
    }
}

File: ios/PepoMote/Net/ScreenClient.swift
import Foundation
import Network
import UIKit

/// Canal de pantalla del GamePad (PROTOCOL.md §4.4): conexión TCP propia al
/// puerto PMP del receptor, línea `screen` con la sesión y el tamaño máximo
/// que cabe en la zona táctil, y de vuelta tramas `ScreenFrames` con la
/// pantalla del GamePad de Cemu en JPEG. Todo en una cola propia: ahí se
/// decodifica y se publica en `image` (cola principal); la UI avisa con
/// `shown` al pintarla y entonces sale el byte de confirmación (una imagen en
/// vuelo: la latencia no se acumula). Si nadie la pinta en `shownFallbackMs`,
/// se confirma igualmente.
///
/// - Sin respuesta al `screen` en 3 s (o cierre sin contestar, o `err`):
///   «sin pantalla» (receptor sin canal de pantalla); reintento cada 5 s.
/// - Sin datos en 6 s tras el ok: conexión muerta; se cierra y se reconecta
///   al segundo, y si tampoco, cada 5 s.
/// - `close` es definitivo: nada vuelve a abrirse ni a publicarse.
final class ScreenClient: ObservableObject {
    /// Resolución nativa de la pantalla del GamePad: nunca se pide más.
    static let nativeWidth = 854
    static let nativeHeight = 480
    static let quality = 70

    /// Estados centinela: la pantalla los traduce (`GamePadScreen`).
    static let statusUnavailable = "\u{0}unavailable"
    static let statusReconnecting = "\u{0}reconnecting"
    /// Prefijo + detalle que manda el PC («sesión desconocida»…).
    static let statusDetail = "\u{0}detail:"

    /// Imagen decodificada lista para pintar; `seq` la identifica.
    struct Image {
        let seq: Int
        let image: UIImage
    }

    @Published private(set) var image: Image?
    /// Última línea de estado (del receptor, o propia: sin pantalla / reconectando).
    @Published private(set) var status: String?
    /// Imágenes por segundo (media de 1 s); 0 sin imagen.
    @Published private(set) var fps: Int = 0

    private let host: String
    private let port: Int
    private let sessionId: UInt32
    private let maxWidth: Int
    private let maxHeight: Int
    private let quality: Int
    private let replyTimeoutMs: Int
    private let dataTimeoutMs: Int
    private let shownFallbackMs: Int
    private let retryAfterDropMs: Int
    private let retryUnavailableMs: Int
    private let decode: (Data) -> UIImage?

    private let q = DispatchQueue(label: "pepomote.screen")
    private var conn: NWConnection?
    private var frames = ScreenFrames()
    private var running = true
    private var started = false
    /// Generación de la conexión: una confirmación de la anterior no se escribe en la nueva.
    private var gen = 0
    private var gotOk = false
    private var seq = 0
    private var pendingAck = 0
    private var replyTimer: DispatchWorkItem?
    private var dataTimer: DispatchWorkItem?
    private var retryTimer: DispatchWorkItem?
    private var fpsWindowStartNs: UInt64 = 0
    private var fpsCount = 0

    init(
        host: String,
        port: Int,
        sessionId: UInt32,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int = ScreenClient.quality,
        replyTimeoutMs: Int = 3000,
        dataTimeoutMs: Int = 6000,
        shownFallbackMs: Int = 200,
        retryAfterDropMs: Int = 1000,
        retryUnavailableMs: Int = 5000,
        decode: @escaping (Data) -> UIImage? = { UIImage(data: $0) }
    ) {
        self.host = host
        self.port = port
        self.sessionId = sessionId
        self.maxWidth = maxWidth
        self.maxHeight = maxHeight
        self.quality = quality
        self.replyTimeoutMs = replyTimeoutMs
        self.dataTimeoutMs = dataTimeoutMs
        self.shownFallbackMs = shownFallbackMs
        self.retryAfterDropMs = retryAfterDropMs
        self.retryUnavailableMs = retryUnavailableMs
        self.decode = decode
    }

    // MARK: - Línea de respuesta (puro, testeable)

    /// Línea `screen` (PROTOCOL.md §4.4); `session_id` va como u32.
    static func helloLine(sessionId: UInt32, w: Int, h: Int, q: Int) -> String {
        "{\"m\":\"screen\",\"session_id\":\(sessionId),\"w\":\(w),\"h\":\(h),\"q\":\(q)}"
    }

    /// `{"m":"screen","ok":true}`: el receptor acepta.
    static func replyOk(_ line: String) -> Bool {
        jsonString(line, "m") == "screen" && line.range(of: "\"ok\"\\s*:\\s*true", options: .regularExpression) != nil
    }

    /// Valor de una clave string de la línea JSON de respuesta (solo tiene
    /// `m`, `ok`, `code` y `msg`), con los escapes deshechos.
    static func jsonString(_ line: String, _ key: String) -> String? {
        let pattern = "\"" + NSRegularExpression.escapedPattern(for: key) + "\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\""
        guard let re = try? NSRegularExpression(pattern: pattern),
              let m = re.firstMatch(in: line, range: NSRange(line.startIndex..., in: line)),
              let r = Range(m.range(at: 1), in: line)
        else { return nil }
        let raw = String(line[r])
        if !raw.contains("\\") { return raw }
        var out = ""
        let chars = Array(raw)
        var i = 0
        while i < chars.count {
            let c = chars[i]
            if c != "\\" || i + 1 >= chars.count {
                out.append(c)
                i += 1
                continue
            }
            let e = chars[i + 1]
            switch e {
            case "n": out.append("\n")
            case "t": out.append("\t")
            case "r": out.append("\r")
            case "u":
                if i + 5 < chars.count, let code = UInt32(String(chars[(i + 2)...(i + 5)]), radix: 16),
                   let scalar = Unicode.Scalar(code) {
                    out.unicodeScalars.append(scalar)
                    i += 4
                }
            default: out.append(e) // \" \\ \/
            }
            i += 2
        }
        return out
    }

    // MARK: - Ciclo de vida

    func start() {
        q.async {
            if self.started || !self.running { return }
            self.started = true
            self.connect()
        }
    }

    /// La UI acaba de pintar la imagen `seq`: ahora sí, confirmación al receptor.
    func shown(_ seq: Int) {
        q.async {
            if self.pendingAck == seq, seq != 0 {
                self.pendingAck = 0
                self.writeAck(self.gen)
            }
        }
    }

    func close() {
        q.async {
            self.running = false
            self.retryTimer?.cancel()
            self.replyTimer?.cancel()
            self.dataTimer?.cancel()
            self.conn?.cancel()
            self.conn = nil
            DispatchQueue.main.async {
                self.image = nil
                self.status = nil
                self.fps = 0
            }
        }
    }

    // MARK: - Una conexión (todo en `q`)

    private func connect() {
        guard running else { return }
        gen += 1
        let myGen = gen
        gotOk = false
        pendingAck = 0
        frames.reset()
        let tcp = NWProtocolTCP.Options()
        tcp.noDelay = true
        tcp.connectionTimeout = 4
        let params = NWParameters(tls: nil, tcp: tcp)
        let nwPort = NWEndpoint.Port(rawValue: UInt16(clamping: port)) ?? NWEndpoint.Port(rawValue: 26761)!
        let c = NWConnection(host: NWEndpoint.Host(host), port: nwPort, using: params)
        conn = c
        c.stateUpdateHandler = { [weak self] state in
            guard let self, self.gen == myGen else { return }
            switch state {
            case .ready:
                let line = ScreenClient.helloLine(sessionId: self.sessionId, w: self.maxWidth, h: self.maxHeight, q: self.quality) + "\n"
                c.send(content: Data(line.utf8), completion: .contentProcessed { _ in })
                self.armReplyTimer(myGen)
                self.receive(c, myGen)
            case .failed, .cancelled:
                self.ended(myGen)
            default:
                break
            }
        }
        // Sin respuesta al connect: también «sin pantalla»
        armReplyTimer(myGen)
        c.start(queue: q)
    }

    private func receive(_ c: NWConnection, _ myGen: Int) {
        c.receive(minimumIncompleteLength: 1, maximumLength: 256 << 10) { [weak self] data, _, isComplete, error in
            guard let self, self.gen == myGen, self.running else { return }
            if let data, !data.isEmpty {
                self.frames.push(data)
                self.consume(myGen)
            }
            if error != nil || isComplete {
                self.ended(myGen)
                return
            }
            self.receive(c, myGen)
        }
    }

    private func consume(_ myGen: Int) {
        if !gotOk {
            // Respuesta: una línea JSON. Un receptor sin pantalla cierra o calla y un `err` trae el motivo.
            guard let line = frames.takeLine() else {
                if frames.pending > 4096 { unavailable(nil); dropConnection(myGen, retryMs: retryUnavailableMs) }
                return
            }
            if !ScreenClient.replyOk(line) {
                unavailable(ScreenClient.jsonString(line, "msg"))
                dropConnection(myGen, retryMs: retryUnavailableMs)
                return
            }
            gotOk = true
            replyTimer?.cancel()
            DispatchQueue.main.async { self.status = nil }
        }
        armDataTimer(myGen)
        while let f = frames.next() {
            switch f.type {
            case ScreenFrames.typeImage: onImage(f.payload, myGen)
            case ScreenFrames.typeStatus: onStatus(String(decoding: f.payload, as: UTF8.self))
            default: break // keepalive (ya contó como dato al llegar) o tipo futuro
            }
        }
    }

    private func onImage(_ data: Data, _ myGen: Int) {
        guard running else { return }
        guard let img = decode(data) else {
            writeAck(myGen) // nada que enseñar: se confirma igual para que siga el flujo
            return
        }
        seq += 1
        let n = seq
        pendingAck = n
        countImage()
        DispatchQueue.main.async { self.image = Image(seq: n, image: img) }
        // Si la UI no llega a pintarla, que no se atasque el flujo
        q.asyncAfter(deadline: .now() + .milliseconds(shownFallbackMs)) { [weak self] in
            guard let self else { return }
            if self.pendingAck == n {
                self.pendingAck = 0
                self.writeAck(myGen)
            }
        }
    }

    /// Estado del receptor: solo lo manda cuando no tiene imagen que dar.
    private func onStatus(_ text: String) {
        resetFps()
        DispatchQueue.main.async {
            self.image = nil
            self.status = text
        }
    }

    private func unavailable(_ msg: String?) {
        guard running else { return }
        resetFps()
        let s = (msg ?? "").isEmpty ? ScreenClient.statusUnavailable : ScreenClient.statusDetail + (msg ?? "")
        DispatchQueue.main.async {
            self.image = nil
            self.status = s
        }
    }

    private func dropped() {
        guard running else { return }
        resetFps()
        DispatchQueue.main.async {
            self.image = nil
            self.status = ScreenClient.statusReconnecting
        }
    }

    /// La conexión ha terminado (fallo, cierre del PC, timeout): según la fase,
    /// «sin pantalla» o «reconectando», y reintento.
    private func ended(_ myGen: Int) {
        guard gen == myGen else { return }
        if gotOk {
            dropped()
            dropConnection(myGen, retryMs: retryAfterDropMs)
        } else {
            unavailable(nil)
            dropConnection(myGen, retryMs: retryUnavailableMs)
        }
    }

    private func dropConnection(_ myGen: Int, retryMs: Int) {
        guard gen == myGen else { return }
        gen += 1 // invalida los callbacks de esta conexión
        replyTimer?.cancel()
        dataTimer?.cancel()
        conn?.cancel()
        conn = nil
        guard running else { return }
        let w = DispatchWorkItem { [weak self] in self?.connect() }
        retryTimer = w
        q.asyncAfter(deadline: .now() + .milliseconds(retryMs), execute: w)
    }

    private func armReplyTimer(_ myGen: Int) {
        replyTimer?.cancel()
        let w = DispatchWorkItem { [weak self] in
            guard let self, self.gen == myGen, !self.gotOk else { return }
            self.unavailable(nil)
            self.dropConnection(myGen, retryMs: self.retryUnavailableMs)
        }
        replyTimer = w
        q.asyncAfter(deadline: .now() + .milliseconds(replyTimeoutMs), execute: w)
    }

    private func armDataTimer(_ myGen: Int) {
        dataTimer?.cancel()
        let w = DispatchWorkItem { [weak self] in
            guard let self, self.gen == myGen else { return }
            self.ended(myGen)
        }
        dataTimer = w
        q.asyncAfter(deadline: .now() + .milliseconds(dataTimeoutMs), execute: w)
    }

    /// Solo en `q`.
    private func writeAck(_ myGen: Int) {
        guard myGen == gen, let c = conn else { return }
        c.send(content: Data([ScreenFrames.ack]), completion: .contentProcessed { _ in })
    }

    private func countImage() {
        let now = DispatchTime.now().uptimeNanoseconds
        if fpsWindowStartNs == 0 { fpsWindowStartNs = now }
        fpsCount += 1
        let elapsed = now - fpsWindowStartNs
        if elapsed >= 1_000_000_000 {
            let f = Int((Double(fpsCount) * 1e9 / Double(elapsed)).rounded())
            DispatchQueue.main.async { self.fps = f }
            fpsWindowStartNs = now
            fpsCount = 0
        }
    }

    private func resetFps() {
        fpsWindowStartNs = 0
        fpsCount = 0
        DispatchQueue.main.async { self.fps = 0 }
    }
}
